### PR TITLE
Refund order based on id or reference number

### PR DESCRIPTION
**Pull request by annagav**
- [ ] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fix https://github.com/mitodl/mitxonline/issues/816

#### What's this PR do?
Refund order based on id or reference number

#### How should this be manually tested?
Try to refund an order through django admin, and through google sheets

### CODE REVIEW

File: ecommerce/api.py [hunk 1]
```diff
@@ -224,12 +224,13 @@ def establish_basket(request):
     return basket
 
 
-def refund_order(*, order_id: int, **kwargs):
+def refund_order(*, order_id: int = None, reference_number=None, **kwargs):
```
Comment (rhysyngsun) [inline ecommerce/api.py:227]: Could you type `reference_number` as a `str` in the signature?

File: ecommerce/api.py [hunk 2]
```diff
@@ -242,18 +243,26 @@ def refund_order(*, order_id: int, **kwargs):
...
-
+        if reference_number is not None:
+            order = FulfilledOrder.objects.select_for_update().get(
+                reference_number=reference_number
+            )
+        elif order_id is not None:
+            order = FulfilledOrder.objects.select_for_update().get(pk=order_id)
+        else:
+            log.debug(
+                f"Either order_id or reference_number is required to fetch the Order. "
```
Comment (rhysyngsun) [inline ecommerce/api.py:254]: Doesn't need to be an f-string and has an extra trailing space. This should also probably be an error.

File: ecommerce/api.py [hunk 3]
```diff
@@ -242,18 +243,26 @@ def refund_order(*, order_id: int, **kwargs):
...
+            )
+            return False
         if order.state != Order.STATE.FULFILLED:
-            log.debug(f"Order with order_id {order_id} is not in fulfilled state.")
+            log.debug(f"Order with order_id {order.id} is not in fulfilled state.")
             return False
 
-        order_recent_transaction = Transaction.objects.filter(order=order_id).first()
+        order_recent_transaction = Transaction.objects.filter(order=order.id).first()
```
Comment (rhysyngsun) [inline ecommerce/api.py:261]: Nit: could be `order.transactions.first()`

File: ecommerce/api.py [hunk 4]
```diff
@@ -242,18 +243,26 @@ def refund_order(*, order_id: int, **kwargs):
...
+        elif order_id is not None:
+            order = FulfilledOrder.objects.select_for_update().get(pk=order_id)
+        else:
+            log.debug(
+                f"Either order_id or reference_number is required to fetch the Order. "
+            )
+            return False
         if order.state != Order.STATE.FULFILLED:
-            log.debug(f"Order with order_id {order_id} is not in fulfilled state.")
+            log.debug(f"Order with order_id {order.id} is not in fulfilled state.")
```
Comment (rhysyngsun) [inline ecommerce/api.py:258]: For this and the other log strings, it should use `%s` formatting so they get deduplicated in sentry.

File: ecommerce/api_test.py [hunk 5]
```diff
@@ -74,6 +74,15 @@ def test_cybersource_refund_no_fulfilled_order(order_state):
     assert refund_response is False
 
 
+def test_cybersource_refund_no_order_id():
+    """Test that refund_order returns logs properly and False when there is no Fulfilled order against
+    the given order_id"""
+
+    refund_response = refund_order()
+    assert f"Either order_id or reference_number is required to fetch the Order."
```
Comment (rhysyngsun) [inline ecommerce/api_test.py:82]: Doesn't need to be an f-string.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
Code review, will do a functional test tomorrow morning.

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
I'm getting this error trying to test, which I think needs to be fixed in the refunds library which is trying to parse the order id column as an int still:

```
[2022-08-17 22:41:01] ERROR 7 [mitol.google_sheets_refunds.api] api.py:61 - [d28c18aed9c6] - Parsing failure
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/mitol/google_sheets_refunds/utils.py", line 79, in parse_raw_data
    order_id=int(raw_row_data[6]),
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/mitol/google_sheets_refunds/api.py", line 59, in process_row
    refund_req_row = RefundRequestRow.parse_raw_data(row_index, row_data)
  File "/usr/local/lib/python3.9/site-packages/mitol/google_sheets_refunds/utils.py", line 95, in parse_raw_data
    raise SheetRowParsingException(str(exc)) from exc
mitol.google_sheets.exceptions.SheetRowParsingException: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

**Comment by annagav** [comment]
closed by accident

**Comment by rhysyngsun** [comment]
@annagav I've cut a new release of mitol-django-common that should fix the test failure: https://pypi.org/project/mitol-django-common/2.5.2/

**Comment by rhysyngsun** [comment]
@annagav I've been trying to test this but I'm finding that it's failing because my google sheet isn't copying over reference numbers - it's still copying over historical numeric order ids but not the ones with the prefix. Any ideas on why or what I need to fix?

File: sheets/plugins.py [hunk 6]
```diff
@@ -12,7 +12,7 @@ def refunds_process_request(
     ) -> RefundResult:
 
         refund_api_success = refund_order(
-            order_id=refund_request_row.order_id, unenroll=True
+            reference_number=refund_request_row.order_id, unenroll=True
```
Comment (rhysyngsun) [inline sheets/plugins.py:15]: Needs to be fixed to use the new property:

```suggestion
            reference_number=refund_request_row.order_ref_num, unenroll=True
```

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
I don't think you'd meant to change the other two files in your last commit.

**Review by rhysyngsun — APPROVED** [review]
:+1: when tests pass

**Pull request merged**